Add NativePromise::AutoRejectProducer object

jyavenard opened the pull request:
#### 1dff68a12c0366f37f3a93dfe94680f0c303d9d3
<pre>
Add NativePromise::AutoRejectProducer object
<a href="https://bugs.webkit.org/show_bug.cgi?id=269185">https://bugs.webkit.org/show_bug.cgi?id=269185</a>
<a href="https://rdar.apple.com/122790738">rdar://122790738</a>

Reviewed by Youenn Fablet.

An AutoRejectProducer is a NativePromise::Producer that will automatically
reject its promise if it&apos;s not been settled at the time it gets deleted.

Add API tests.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/274535@main">https://commits.webkit.org/274535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d12c8a72ae179ea2cc82e9308d0ea016ced7dea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41640 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41821 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35187 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41593 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15595 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32870 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15374 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34051 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13366 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/39153 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13327 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34991 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43099 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/32740 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35669 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35320 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39134 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/38913 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11633 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37374 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15705 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/45919 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8803 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9363 "Found unexpected failure with change (failure)") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15327 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1244,8 +1246,10 @@ class NativePromiseProducer final : public ConvertibleToNativePromise {
     // used by IsConvertibleToNativePromise to determine how to cast the result.
-    using PromiseType = NativePromise<ResolveValueT, RejectValueT, options>;
+    using PromiseType = NativePromise<ResolveValueT, RejectValueT, options & ~static_cast<unsigned>(PromiseOption::AutoRejectProducer)>;
+    static constexpr bool AutoReject = options & PromiseOption::AutoRejectProducer;
```
Suggestion: Add `static constexpr bool AutoRejectNonVoid = AutoReject && !std::is_void_v<RejectValueT>;`, I think it would clarify 4 expressions below, especially the first constructor where you're negating it.

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1385,6 +1409,12 @@ class NativePromiseProducer final : public ConvertibleToNativePromise {
     }
 
+    template<typename RejectValueType_, typename = std::enable_if<AutoRejectNonVoid>>
+    void setDefaultReject(RejectValueType_&& rejectValue)
```
Why do we need this setter?

jyavenard replied:
I felt a good things to have if along the lifecycle of the producer we ever wanted to change the reject value

jyavenard committed:
Add NativePromise::AutoRejectProducer object
https://bugs.webkit.org/show_bug.cgi?id=269185
rdar://122790738

Reviewed by Youenn Fablet.

An AutoRejectProducer is a NativePromise::Producer that will automatically
reject its promise if it's not been settled at the time it gets deleted.

Add API tests.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/274535@main
